salt/modules/shadow.py

```diff
@@ -306,81 +392,145 @@ def set_password(name, password, use_usermod=False):
-        uinfo = info(name)
+        uinfo = info(name, root=root)
         return uinfo['passwd'] == password
 
 
-def set_warndays(name, warndays):
+def set_warndays(name, warndays, root=None):
     '''
     Set the number of days of warning before a password change is required.
     See man chage.
 
+    name
+        User to modify
+
+    warndays
+        Number of days of warning before a password change is required
+
+    root
+        Directory to chroot into
+
     CLI Example:
 
     .. code-block:: bash
 
         salt '*' shadow.set_warndays username 7
     '''
-    pre_info = info(name)
-    if warndays == pre_info['warn']:
-        return True
-    cmd = 'chage -W {0} {1}'.format(warndays, name)
-    __salt__['cmd.run'](cmd, python_shell=False)
-    post_info = info(name)
-    if post_info['warn'] != pre_info['warn']:
-        return post_info['warn'] == warndays
-    return False
+    return _set_attrib(name, 'warn', warndays, '-W', root=root)
 
 
-def set_date(name, date):
+def set_date(name, date, root=None):
     '''
     Sets the value for the date the password was last changed to days since the
     epoch (January 1, 1970). See man chage.
 
+    name
+        User to modify
+
+    date
+        Date the password was last changed
+
+    root
+        Directory to chroot into
+
     CLI Example:
 
     .. code-block:: bash
 
         salt '*' shadow.set_date username 0
     '''
-    cmd = 'chage -d {0} {1}'.format(date, name)
-    return not __salt__['cmd.run'](cmd, python_shell=False)
+    return _set_attrib(name, 'lstchg', date, '-d', root=root, validate=False)
 
 
-def set_expire(name, expire):
+def set_expire(name, expire, root=None):
     '''
     .. versionchanged:: 2014.7.0
 
     Sets the value for the date the account expires as days since the epoch
     (January 1, 1970). Using a value of -1 will clear expiration. See man
     chage.
 
+    name
+        User to modify
+
+    date
+        Date the account expires
+
+    root
+        Directory to chroot into
+
     CLI Example:
 
     .. code-block:: bash
 
         salt '*' shadow.set_expire username -1
     '''
-    cmd = 'chage -E {0} {1}'.format(expire, name)
-    return not __salt__['cmd.run'](cmd, python_shell=False)
+    return _set_attrib(name, 'expire', expire, '-E', root=root, validate=False)
 
 
-def list_users():
+def list_users(root=None):
     '''
     .. versionadded:: 2018.3.0
 
     Return a list of all shadow users
 
+    root
+        Directory to chroot into
+
     CLI Example:
 
     .. code-block:: bash
 
         salt '*' shadow.list_users
     '''
-    return sorted([user.sp_nam for user in spwd.getspall()])
+    if root is not None:
+        _fun = functools.partial(_getspall, root=root)
+    else:
+        _fun = functools.partial(spwd.getspall)
+
+    return sorted([user.sp_namp if hasattr(user, 'sp_namp') else user.sp_nam
+                   for user in _fun()])
+
+
+def _getspnam(name, root=None):
```

Comment:
This looks somewhat very similar to what is used in other module. You could join all that and make utility function instead, removing lots of code duplication from both places.

Reply:
Ah indeed I was tempted to do this. Actually this patch removes a lot of old duplicated code. But in this case the interpretation of the columns in different for user, groups and shadow. The only generic part is open the file, read each line, and split the line with ':'

I will think what I can factor out.